### PR TITLE
Publish APM metrics from the Azure BlobStore

### LINKED ISSUE #104443: Request stats should track retries for Azure storage client
Azure storage client has a default of 3 [max_retries](https://www.elastic.co/guide/en/elasticsearch/reference/current/repository-azure.html). The number of retries can be retrieved from the `Context` object under the key `HttpLoggingPolicy. RETRY_COUNT_CONTEXT`. Today, when tracking number of requests on the client side, we [do _not_ count the retries](https://github.com/elastic/elasticsearch/blob/d1dfeba31390cb02191d1d82610595fd7daa9951/modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java#L255-L261). This means we can underestimate the number of requests that server has processed. This has led to CI failure (#104362). This behaviour is also inconsistent with how we track s3 storage requests where we do include retries. We should consider tracking retries for Azure storage request stats.

### PR DESCRIPTION
**Pull request by nicktindall**
Instruments the `AzureRepository` with APM metrics.

I attempted to make the metrics equivalent to what would be published by the S3 repository under the same circumstances. Specifically;

| metric name | description |
|-----------------------------------------------------|------------|
|`es.repositories.requests.total`| Is incremented for each request sent to the blob store (including retries)|
|`es.repositories.exceptions.total`| Is incremented for each request which returns a non `2xx` response OR fails to return a response (includes throttling and retryable errors)|
|`es.repositories.exceptions.request_range_not_satisfied.total`| Is incremented each time an operation ends with a `416` response|
|`es.repositories.throttles.total`| Is incremented each time we receive a `429` response|
|`es.repositories.operations.total`| Is incremented for each operation we attempt, whether it succeds or fails, this doesn't include retries|
|`es.repositories.operations.unsuccessful.total`| Is incremented for each operation that ends with a non `2xx` response or throws an exception|
|`es.repositories.exceptions.histogram`| Each time an operation has one or more failed requests (from non `2xx` response or exception), the count of those is sampled|
|`es.repositories.throttles.histogram`| Each time an operation has one or more throttled requests, the count of those is sampled|
|`es.repositories.requests.http_request_time.histogram`| Any operation that returns a response will record its request time. Operations that never receive a response from the blob store do not record a value. The time recorded is from before the first request is sent until after the last response is received (i.e. it includes any retries). Doesn't include the consumption of the body of the response.|

As an example:

I call `BlobStore#getInputStream` to read a blob from the blob store
- counts as one `es.repositories.operations.total`
- We send GET /account/container/{blobPath} 
  - counts as one `es.repositories.requests.total`
- We get back a 429 response
  - counts as one `es.repositories.throttles.total`
  - counts as one `es.repositories.exceptions.total`
- We re-send the GET
  - counts as one `es.repositories.requests.total`
- We get back a 500 response 
  - counts as one `es.repositories.exceptions.total`
- We re-send the GET
  - counts as one `es.repositories.requests.total`
- We get back a 416 response
  - counts as one `es.repositories.exceptions.total`
  - counts as one `es.repositories.exceptions.request_range_not_satisfied.total`
- We complete the operation with a 416 response
  - counts as one `es.repositories.operations.unsuccessful.total`

So the metrics for that blob retrieval are:
- add 1 to `es.repositories.operations.total`
- add 3 to `es.repositories.requests.total`
- add 3 to `es.repositories.exceptions.total`
- add 1 to `es.repositories.throttles.total`
- add 1 to `es.repositories.exceptions.request_range_not_satisfied.total`
- add 1 to `es.repositories.operations.unsuccessful.total`
- record 3 against `es.repositories.exceptions.histogram`
- record 1 against `es.repositories.throttles.histogram`
- record time for the whole flow in milliseconds against `es.repositories.requests.http_request_time.histogram`

Closes [ES-9550](https://elasticco.atlassian.net/browse/ES-9550)

### CODE REVIEW

File: modules/repository-azure/src/internalClusterTest/java/org/elasticsearch/repositories/azure/AzureBlobStoreRepositoryTests.java [hunk 1]
```diff
@@ -195,24 +219,13 @@ protected String requestUniqueId(final HttpExchange exchange) {
...
 
         @Override
         protected void maybeTrack(String request, Headers headers) {
-            // Same request id is a retry
-            // https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-ncnbi/817da997-30d2-4cd3-972f-a0073e4e98f7
-            // Do not count retries since the client side request stats do not track them yet.
-            // See https://github.com/elastic/elasticsearch/issues/104443
-            if (false == seenRequestIds.add(headers.getFirst("X-ms-client-request-id"))) {
-                return;
-            }
```
Comment (nicktindall) [inline modules/repository-azure/src/internalClusterTest/java/org/elasticsearch/repositories/azure/AzureBlobStoreRepositoryTests.java:215]: This was here because of #104443

File: modules/repository-azure/src/internalClusterTest/java/org/elasticsearch/repositories/azure/AzureBlobStoreRepositoryTests.java [hunk 2]
```diff
@@ -303,4 +316,91 @@ public void testReadByteByByte() throws Exception {
...
+                    metric.getLong(),
+                    equalTo(statsCollectors.get(statsKey).counter.sum())
+                );
+                aggregatedMetrics.compute(statsKey.operation(), (k, v) -> v == null ? metric.getLong() : v + metric.getLong());
+            });
+        }
+
+        // Metrics number should be consistent with server side request count as well.
+        assertThat(aggregatedMetrics, equalTo(getServerMetrics()));
+    }
```
Comment (nicktindall) [inline modules/repository-azure/src/internalClusterTest/java/org/elasticsearch/repositories/azure/AzureBlobStoreRepositoryTests.java:399]: This is the same as the test with the same name for the S3 repository, but we can't test operational purpose here because we don't convey that to the server in Azure.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java [hunk 3]
```diff
@@ -189,8 +193,9 @@ AzureBlobServiceClient createClient(
             builder.credential(credentialBuilder.build());
         }
 
-        if (successfulRequestHandler != null) {
-            builder.addPolicy(new SuccessfulRequestTracker(purpose, successfulRequestHandler));
+        if (requestMetricsHandler != null) {
+            builder.addPolicy(new RequestMetricsTracker(purpose, requestMetricsHandler));
+            builder.addPolicy(RetryMetricsTracker.INSTANCE);
```
Comment (nicktindall) [inline modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:198]: `RequestMetricsTracker` sets up the context at the beginning of the request and dispatches the metrics to the listener at the end of the request, `RetryMetricsTracker` bumps the necessary counters for each retry (it's stateless and hence reusable).

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java [hunk 4]
```diff
@@ -259,38 +264,136 @@ protected void doStop() {
...
+        INSTANCE;
+
+        @Override
+        public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+            RequestMetrics metrics = (RequestMetrics) context.getData(RequestMetricsTracker.ES_REQUEST_METRICS_CONTEXT_KEY)
+                .orElseGet(() -> {
+                    String errorMessage = "No metrics object associated with request " + context.getHttpRequest();
+                    assert false : errorMessage;
+                    throw new IllegalStateException(errorMessage);
+                });
```
Comment (nicktindall) [inline modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:324]: Not sure if this is too heavy-handed? if it occurs its a bug and should never have got through CI.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java [hunk 5]
```diff
@@ -259,38 +264,136 @@ protected void doStop() {
...
     protected void doClose() {}
 
-    private static final class SuccessfulRequestTracker implements HttpPipelinePolicy {
-        private static final Logger logger = LogManager.getLogger(SuccessfulRequestTracker.class);
+    static class RequestMetrics {
+        private volatile long timeToResponseInMillis = -1;
+        private volatile int requestCount;
+        private volatile int errorCount;
+        private volatile int throttleCount;
+        private volatile int statusCode;
```
Comment (nicktindall) [inline modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:272]: These are volatile because they're accessed by multiple threads, but we increment them non-atomically because they should only be modified by one thread at a time.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java [hunk 6]
```diff
@@ -259,38 +264,136 @@ protected void doStop() {
...
+        }
+
+        @Override
+        public HttpPipelinePosition getPipelinePosition() {
+            return HttpPipelinePosition.PER_RETRY;
+        }
+
+        private static boolean isErrorCode(int statusCode) {
+            return statusCode <= 199 || statusCode > 299;
+        }
```
Comment (nicktindall) [inline modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:346]: This will record an "error" for e.g. a 404 that we get when checking if a blob exists. Technically that's probably not an "error", but it seems consistent with the S3 implementation.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java [hunk 7]
```diff
@@ -259,38 +264,136 @@ protected void doStop() {
...
+                logger.debug("Detected error in RequestMetricsTracker", throwable);
+                trackCompletedRequest(context.getHttpRequest(), requestMetrics);
+            });
         }
 
-        private void trackSuccessfulRequest(HttpRequest httpRequest, HttpResponse httpResponse) {
+        private void trackCompletedRequest(HttpRequest httpRequest, RequestMetrics requestMetrics) {
             HttpMethod method = httpRequest.getHttpMethod();
-            if (httpResponse != null && method != null && httpResponse.getStatusCode() > 199 && httpResponse.getStatusCode() <= 299) {
+            if (method != null) {
```
Comment (nicktindall) [inline modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:377]: I'm not sure why this null check is here, but I left it intact just incase

File: server/src/main/java/org/elasticsearch/repositories/RepositoriesMetrics.java [hunk 8]
```diff
@@ -61,4 +65,25 @@ public RepositoriesMetrics(MeterRegistry meterRegistry) {
...
+            "repo_type",
+            repositoryMetadata.type(),
+            "repo_name",
+            repositoryMetadata.name(),
+            "operation",
+            operation,
+            "purpose",
+            purpose.getKey()
+        );
+    }
```
Comment (nicktindall) [inline server/src/main/java/org/elasticsearch/repositories/RepositoriesMetrics.java:87]: It seemed like it'd be a good thing to define this in one place for consistency between repository implementations?

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java [hunk 9]
```diff
@@ -691,40 +687,103 @@ public String getKey() {
...
+                repositoriesMetrics.requestRangeNotSatisfiedExceptionCounter().incrementBy(1, attributes);
+            }
+
+            repositoriesMetrics.operationCounter().incrementBy(1, attributes);
+            if (requestMetrics.getStatusCode() <= 199 || requestMetrics.getStatusCode() > 299) {
+                repositoriesMetrics.unsuccessfulOperationCounter().incrementBy(1, attributes);
+            }
+
+            repositoriesMetrics.requestCounter().incrementBy(requestMetrics.getRequestCount(), attributes);
+            if (requestMetrics.getErrorCount() > 0) {
```
Comment (nicktindall) [inline modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:761]: I wonder if we're losing information by not sampling when the error/throttling count is zero. This is consistent with the S3 metrics but I feel it makes it hard to distinguish between an increase in the error rate and an increase in the volume of requests?

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java (hunk 4 above)
Comment (nicktindall) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java]: Removed the exception and left the assertion. That should be sufficient.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java [hunk 10]
```diff
@@ -259,38 +265,136 @@ protected void doStop() {
...
+    private enum RetryMetricsTracker implements HttpPipelinePolicy {
+        INSTANCE;
+
+        @Override
+        public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+            Optional<Object> metricsData = context.getData(RequestMetricsTracker.ES_REQUEST_METRICS_CONTEXT_KEY);
+            if (metricsData.isPresent() == false) {
+                assert false : "No metrics object associated with request " + context.getHttpRequest();
+                return next.process();
+            }
```
Comment (nicktindall) [inline modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:324]: This would be a bug if this handler were in the pipeline and the metrics wasn't populated in the context, but I made it an assertion because it should be catchable with tests.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java [hunk 11]
```diff
@@ -691,40 +687,103 @@ public String getKey() {
...
+            this.counter = new LongAdder();
+            this.attributes = RepositoriesMetrics.createAttributesMap(repositoryMetadata, purpose, operation.getKey());
+        }
+
+        public void onRequestComplete(AzureClientProvider.RequestMetrics requestMetrics) {
+            counter.add(requestMetrics.getRequestCount());
+
+            // range not satisfied is not retried, so we count them by checking the final response
+            if (requestMetrics.getStatusCode() == RestStatus.REQUESTED_RANGE_NOT_SATISFIED.getStatus()) {
+                repositoriesMetrics.requestRangeNotSatisfiedExceptionCounter().incrementBy(1, attributes);
```
Comment (mhl-b) [inline modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:752]: Does it make sense for operation to contain own subset of request metrics? And then repositoryMetrics is an aggregate of operation metrics. I think there few benefits: remove side-effect of mutating repositoriesMetrics, less contention, gives more granularity in metrics operation/request dimensions. The downside is more code to update operation metrics and repository metrics.

Comment (nicktindall) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:752]: I'm not sure I follow, `RepositoriesMetrics` is a singleton record that registers the APM metrics and provides access to them for places where they're recorded.
`RequestMetrics` in this place is an object containing the metrics for a single operation, `RequestMetrics` is populated over the course of executing the operation, then passed to the `RequestMetricsHandler` at the end of the request, which in this case interprets them to publish via `RepositoriesMetrics`.

I think that singleton pattern is how otel metrics are commonly used (similar to a logger) in the java SDK at least?

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java [hunk 12]
```diff
@@ -102,59 +102,55 @@ public class AzureBlobStore implements BlobStore {
...
+    private final RepositoryMetadata repositoryMetadata;
 
     private final String clientName;
     private final String container;
     private final LocationMode locationMode;
     private final ByteSizeValue maxSinglePartUploadSize;
 
     private final StatsCollectors statsCollectors = new StatsCollectors();
-    private final AzureClientProvider.SuccessfulRequestHandler statsConsumer;
+    private final AzureClientProvider.RequestMetricsHandler statsConsumer;
```
Comment (ywangd) [inline modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:115]: Can we rename this field to `reqestMetricsHandler`? The existing name confuses me.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java [hunk 13]
```diff
@@ -691,40 +687,103 @@ public String getKey() {
...
-        public void onSuccessfulRequest(Operation operation, OperationPurpose purpose) {
-            collectors.computeIfAbsent(new StatsKey(operation, purpose), k -> new LongAdder()).increment();
+        public void onRequestComplete(Operation operation, OperationPurpose purpose, AzureClientProvider.RequestMetrics requestMetrics) {
+            collectors.computeIfAbsent(new StatsKey(operation, purpose), k -> new OperationMetrics(operation, purpose))
+                .onRequestComplete(requestMetrics);
         }
     }
 
+    // visible for testing
+    class OperationMetrics {
```
Comment (ywangd) [inline modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:737]: Nit: can we call this something like `OperationMetricsRecorder` (or `StatsCollector` singular) since we also have `RequestMetrics` which is a data class and this one is not?

**Review by ywangd — COMMENTED** [review]
This overall looks good to me. I plan to take a closer look. Meanwhile I had some minor comments.

The PR description has a nice summary on how each metrics should behave. I wonder whether it should be incorporated into `RepositoryMetrics` since ideally we want the definitions to hold for all repositories.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java (hunk 9 above)
Comment (ywangd) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:761]: Maybe. So far it has been relatively easy to differentiate by just eyeballing the `requests` and `exceptions` lines on the visualization. Part of the reason maybe the increase of requests is currently dominated by internal schedulings, such as SLM runs. So they have been quite obvious. When we have more larger projects that bring in more bursty loads, we might need to introduce something like ratio for more precise differentiation. I think it can wait for now.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java (hunk 11 above)
Comment (mhl-b) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:752]: The `OperationMetrics` seems to me as extra abstraction that has side-effect of changing  `RepositoriesMetrics`. I think you can remove it, something like this:

```java
class AzureRepoMetrics {
  RepositoriesMetrics repoMetrics
  Map<StatsKey, LongAdder> opsCounter
  Map<StatsKey, Attributes> opsAttributes 
  
   void addRequestMetrics(Operation operation, OperationPurpose purpose, AzureClientProvider.RequestMetrics requestMetrics) {
  ... // same as onRequestComplete
  }

  RepositoriesMetrics repoMetrics() { return repoMetrics; }
}

...
public void onRequestComplete(Operation operation, OperationPurpose purpose, AzureClientProvider.RequestMetrics requestMetrics) {
  azureRepoMetrics.addRequestMetrics(operation, purpose, requestMetrics);
}
...
```

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java [hunk 14]
```diff
@@ -259,38 +265,136 @@ protected void doStop() {
...
+                        metrics.throttleCount++;
+                    }
+                }
+            });
+        }
+
+        @Override
+        public HttpPipelinePosition getPipelinePosition() {
+            return HttpPipelinePosition.PER_RETRY;
+        }
```
[the remark this answers is not in the document]
Comment (nicktindall) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:343]: Yeah it was just to make it clear, because there are two `HttpPipelinePolicy` instances that interact with/depend on each other, that one policy is operation-scoped and the other is retry-scoped.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java [hunk 15]
```diff
@@ -259,38 +265,136 @@ protected void doStop() {
...
+            RequestMetrics metrics = (RequestMetrics) metricsData.get();
+            metrics.requestCount++;
+            return next.process().doOnError(throwable -> {
+                logger.debug("Detected error in RetryMetricsTracker", throwable);
+                metrics.errorCount++;
+            }).doOnSuccess(response -> {
+                if (isErrorCode(response.getStatusCode())) {
+                    metrics.errorCount++;
+                    if (response.getStatusCode() == RestStatus.TOO_MANY_REQUESTS.getStatus()) {
+                        metrics.throttleCount++;
```
[the remark this answers is not in the document]
Comment (nicktindall) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:334]: I did check that, from what I could see, Azure seems to always throttle with a 429.

See https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/request-limits-and-throttling#error-code, I will add that link in a comment.

There are [retries for more than just throttling exceptions](https://learn.microsoft.com/en-us/azure/architecture/best-practices/retry-service-specific#general-rest-and-retry-guidelines), but I think we're counting specifically throttles in this metric.

[the remark this answers is not in the document]
Comment (nicktindall) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:334]: AWS client contains more logic because [they use AWS error codes as well as response status](https://github.com/aws/aws-sdk-java/blob/864baf847e35b6c987ad08d81738050d329c515c/aws-java-sdk-core/src/main/java/com/amazonaws/retry/RetryUtils.java#L104) to indicate throttling by the look of things

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java [hunk 16]
```diff
@@ -259,38 +265,136 @@ protected void doStop() {
...
+            final RequestMetrics requestMetrics = new RequestMetrics();
+            final long requestStartTimeInMillis = System.currentTimeMillis();
+            context.setData(ES_REQUEST_METRICS_CONTEXT_KEY, requestMetrics);
+            return next.process().doOnSuccess((httpResponse) -> {
+                requestMetrics.statusCode = httpResponse.getStatusCode();
+                requestMetrics.timeToResponseInMillis = System.currentTimeMillis() - requestStartTimeInMillis;
+                trackCompletedRequest(context.getHttpRequest(), requestMetrics);
+            }).doOnError(throwable -> {
+                logger.debug("Detected error in RequestMetricsTracker", throwable);
+                trackCompletedRequest(context.getHttpRequest(), requestMetrics);
```
[the remark this answers is not in the document]
Comment (nicktindall) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:372]: One thing that _might_ manifest in this way is a request timeout, which we probably want to see reflected in the metrics. Basically any time you get a response back from the server, even 5xx responses, we execute `doOnSuccess`. "Success" here just means a request with a response, it's a successful `Mono` not a successful HTTP request. The errors occur when things happen like the connection is dropped.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java [hunk 17]
```diff
@@ -691,40 +687,103 @@ public String getKey() {
...
+        public void onRequestComplete(AzureClientProvider.RequestMetrics requestMetrics) {
+            counter.add(requestMetrics.getRequestCount());
+
+            // range not satisfied is not retried, so we count them by checking the final response
+            if (requestMetrics.getStatusCode() == RestStatus.REQUESTED_RANGE_NOT_SATISFIED.getStatus()) {
+                repositoriesMetrics.requestRangeNotSatisfiedExceptionCounter().incrementBy(1, attributes);
+            }
+
+            repositoriesMetrics.operationCounter().incrementBy(1, attributes);
+            if (requestMetrics.getStatusCode() <= 199 || requestMetrics.getStatusCode() > 299) {
```
[the remark this answers is not in the document]
Comment (nicktindall) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java]: Addressed in 9300906349df120a124716f17fe9a19086393a74. This logic is duplicated all over the codebase. I didn't update any of the other usages of it as part of this PR but would be happy to consolidate on a single implementation as a follow up if you think it's worthwhile.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (nicktindall) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:336]: Added link to docs in 27697f9fcdef570834b6973dda6c04acbedfbc78

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java (hunk 12 above)
Comment (nicktindall) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java]: Addressed in b847370b0248d1b6625a7af477b1d0cec896be2e

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java (hunk 13 above)
Comment (nicktindall) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java]: Updated this in a commit where I addressed @mhl-b's feedback also: f04831e3b071b921e92e359674554d4b80bfa989

**Review by mhl-b — APPROVED** [review]
LGTM

**Comment by nicktindall** [comment]
@ywangd Metric descriptions added in c381e7c0491d60e18075ea275998fc85474606e8

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (ywangd) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:336]: Thanks a lot for checking and the links

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java (hunk 16 above)
[the remark this answers is not in the document]
Comment (nicktindall) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureClientProvider.java:370]: It appears as though S3 client records `HttpRequestTime` even for requests that end in an exception (assuming timeouts, connection failures etc.), see [here](https://github.com/aws/aws-sdk-java/blob/864baf847e35b6c987ad08d81738050d329c515c/aws-java-sdk-core/src/main/java/com/amazonaws/http/AmazonHttpClient.java#L1369-L1403).

The time we record appears to be all of these durations added together, so that would exclude the time taken to pause between retries. I'll make the Azure client consistent with that.

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java [hunk 18]
```diff
@@ -720,11 +732,49 @@ Map<String, Long> statsMap(boolean stateless) {
...
-            collectors.computeIfAbsent(new StatsKey(operation, purpose), k -> new LongAdder()).increment();
+        public void onRequestComplete(Operation operation, OperationPurpose purpose, AzureClientProvider.RequestMetrics requestMetrics) {
+            final StatsKey statsKey = new StatsKey(operation, purpose);
+            final LongAdder counter = opsCounters.computeIfAbsent(statsKey, k -> new LongAdder());
+            final Map<String, Object> attributes = opsAttributes.computeIfAbsent(
+                statsKey,
+                k -> RepositoriesMetrics.createAttributesMap(repositoryMetadata, purpose, operation.getKey())
+            );
+
+            counter.add(requestMetrics.getRequestCount());
```
Comment (ywangd) [inline modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java:743]: I wonder whether we should split this change into a separate PR. This is a slight change in behaviour so feels better to be in its own PR (albeit tiny). Also, I remembered this comment https://github.com/elastic/elasticsearch/pull/104444#discussion_r1455068279 which could use some clarification

File: modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (ywangd) [reply modules/repository-azure/src/main/java/org/elasticsearch/repositories/azure/AzureBlobStore.java]: Sounds good to me.

**Review by ywangd — APPROVED** [review]
LGTM

I suggest we separate the changes for #104443 into its own PR. Thanks!
